### PR TITLE
[Spark] Integrate domain metadata with OPTIMIZE

### PR DESCRIPTION
**Pull request by zedtang**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Resolves https://github.com/delta-io/delta/issues/2451

Integrates domain metadata with OPTIMIZE to get the clustering columns from DomainMetadata and perform clustering. Disallow specifying partition predicates and ZORDER BY for a clustered table.
## How was this patch tested?
Introduce a new test suite `ClusteredTableClusteringSuite` and add test for running optimize on a clustered table.

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala
Comment (dabao521) [inline spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:226]: nit: The current solution looks fine, but we can use `Enum` to make it cleaner

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala
Comment (dabao521) [inline spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:438]: I must miss something, where is the clustering domain metadata commited in this change?

Comment (zedtang) [reply spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:438]: Domain metadata is committed when creating a clustered table [here](https://github.com/delta-io/delta/blob/b4e5d5c4532f9eb94edcbabe63e613619bf87003/spark/src/main/scala/org/apache/spark/sql/delta/commands/CreateDeltaTableCommand.scala#L319).

This PR reads the clustering column from domain metadata using the [extractLogicalNames](https://github.com/delta-io/delta/pull/2461/files#diff-3773dfb6ca40a45d8407d111b056fb5c6acea3153eb59d05e3e0476bfc55a7a3R236) helper.

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala
Comment (imback82) [inline spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:150]: put this under `if (ClusteredTableUtils.isSupported(txn.protocol))`?

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala
Comment (imback82) [inline spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:236]: add an assertion here: `assert(isClusteredTable)`

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala
Comment (imback82) [inline spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:228]: do we need this variable? or directly use the condition below?

File: spark/src/test/scala/org/apache/spark/sql/delta/clustering/ClusteredTableClusteringSuite.scala
Comment (imback82) [inline spark/src/test/scala/org/apache/spark/sql/delta/clustering/ClusteredTableClusteringSuite.scala:63]: can you assert that none of these files are clustered? (no clustering provider tag set)

File: spark/src/test/scala/org/apache/spark/sql/delta/skipping/ClusteredTableTestUtils.scala
Comment (imback82) [inline spark/src/test/scala/org/apache/spark/sql/delta/skipping/ClusteredTableTestUtils.scala:41]:
```suggestion
   * Runs optimize on the table and calls postHook on the metrics.
```

File: spark/src/test/scala/org/apache/spark/sql/delta/clustering/ClusteredTableClusteringSuite.scala
Comment (imback82) [inline spark/src/test/scala/org/apache/spark/sql/delta/clustering/ClusteredTableClusteringSuite.scala:68]: is it possible to check if hilbert was used? (doesn't block this PR)

**Review by imback82 — COMMENTED** [review]
LGTM (pending comments)!

File: spark/src/test/scala/org/apache/spark/sql/delta/clustering/ClusteredTableClusteringSuite.scala
Comment (zedtang) [reply spark/src/test/scala/org/apache/spark/sql/delta/clustering/ClusteredTableClusteringSuite.scala:67]: Done

File: spark/src/test/scala/org/apache/spark/sql/delta/clustering/ClusteredTableClusteringSuite.scala
Comment (zedtang) [reply spark/src/test/scala/org/apache/spark/sql/delta/clustering/ClusteredTableClusteringSuite.scala:73]: Let me do this as a follow-up in https://github.com/delta-io/delta/issues/2449

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala
Comment (zedtang) [reply spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:226]: good suggestion, I'll do it in https://github.com/delta-io/delta/issues/2449

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala
Comment (imback82) [reply spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:150]: yea, but why waste calling it for the non-clustered table scenario?

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala
Comment (imback82) [inline spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:236]: is this a valid scenario? when does this happen?

**Review by imback82 — COMMENTED** [review]
did one more round

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala
Comment (imback82) [reply spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:226]: you may want to update https://github.com/delta-io/delta/issues/2449 to easily track this.

File: spark/src/test/scala/org/apache/spark/sql/delta/clustering/ClusteredTableClusteringSuite.scala
Comment (imback82) [reply spark/src/test/scala/org/apache/spark/sql/delta/clustering/ClusteredTableClusteringSuite.scala:73]: you may want to update https://github.com/delta-io/delta/issues/2449 to easily track this.

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala
Comment (imback82) [reply spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:150]: in terms of calls, it may be fine (you need to run `isSupported` at least once anyway`. But I think it makes more sense to put the validation under the below `if` along with other checks.

Comment (zedtang) [reply spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:150]: Yeah, makes sense. I moved to inside the if below.

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala
Comment (zedtang) [reply spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:226]: Updated the issue, thanks!

File: spark/src/test/scala/org/apache/spark/sql/delta/clustering/ClusteredTableClusteringSuite.scala
Comment (zedtang) [reply spark/src/test/scala/org/apache/spark/sql/delta/clustering/ClusteredTableClusteringSuite.scala:73]: Updated!

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala
Comment (zedtang) [reply spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:236]: This path gets triggered when we call OPTIMIZE on a non-clustered table, so it triggered compaction.

I can't do the assert in https://github.com/delta-io/delta/pull/2461/files#r1446787772 because OPTIMIZE(compaction) will trigger this path and will hit the assert.

Comment (imback82) [reply spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:236]: make it lazy val?

Comment (zedtang) [reply spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala]: Looks like I can't make `clusteringColumns` a lazy val since `getOperation` relies on it being empty for compaction scenario.

So I made `curve` a lazy val and put the assert there.

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaErrors.scala
Comment (imback82) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaErrors.scala:3132]: to be consistent with below
```suggestion
      messageParameters = Array(s"${predicates.mkString(" ")}"))
```

**Review by imback82 — APPROVED** [review]
LGTM!

**Pull request closed without merging**
